uiowa/uiowa
# [Classrooms] Entity usage stuff

Resolves #7675 





# How to test



```
ddev blt ds --site=classrooms.uiowa.edu
```
Go grab some lunch, a coffee, read a book...It'll take awhile.

```
ddev drush @classrooms.local pmu entity_usage && ddev drush @classrooms.local en entity_usage && ddev drush @classrooms.local uli /admin/config/entity-usage/batch-update
```

Bulk update the entity usage table
Use your preferred db tool and check out the the `entity_usage` table, or use
```
ddev drush @classrooms.local sqlq "select count(*) from node_revision" && ddev drush @classrooms.local sqlq "select count(*) from entity_usage"
```
Both should be more reasonable (in the 100s to 1000s for node revisions and under 100k for entity usage)

Run the classrooms rooms import
```
ddev drush @classrooms.local classrooms-rooms
```
No (or very few) updates should be made from it. Check the node revisions and entity usage tables again, and they should be the same size as before.

Delete and re-add a room or edit a room and change the imported values
An example of this could be
```
ddev drush @classrooms.local sqlq "delete from node__field_room_technology_features where entity_id=1066"
```
which would remove all tech features from node 1066, which if the importer works correctly, should be seen as a mismatch with the source and that node (and only that node) should get updated. If there are fewer than 3 revisions for that node, the `node_revision` count should not increase as a result.

Note: The room needs to be updated like this or you need to comment out the code in the room presave, otherwise it will update upon saving and you will still have 0 records which need updating.

## Followup for merging

I added the 'needs scheduled deployment' label. I don't think it's strictly necessary, but the update does take quite a bit of time, and it may be good to avoid possible overlap with site edits.

After deployment, the prod `entity_usage` table should be truncated and regenerated. Probably the easiest way to do this is like the testing instructions here, and to just uninstall and reinstall the module, then log in and bulk update the table.
